src/typechecker.erl

```diff
@@ -1050,7 +1051,7 @@ expect_list_union([], AccTy, AccCs, _NoAny, _N, _TEnv) ->
 infer_literal_string("") ->
     type(nil);
 infer_literal_string(Str) ->
-    SortedChars = lists:usort(Str),
+    SortedChars = ?assert_type(lists:usort(Str), [A, ...]),
```

Comment:
Why type variable `A`? We could use `char()` or just `[any(), ...]` or `nonempty_list()` if we don't care.

```suggestion
    SortedChars = ?assert_type(lists:usort(Str), [char(), ...]),
```

Another idea hit me: If we know by type refinement that it's a non-empty list here, maybe could override the type of lists:usort in the prelude with a spec that preserves the non-emptiness:

```Erlang
-spec usort([A, ...]) -> [A, ...];
           ([])       -> [].

Reply:
Hmm, that's clever, but wouldn't it only make sense with intersection types?
Another thing is that this trick would have to be used with a number of list functions that return an empty list for an empty input list. Still, it only has to be done once for the standard library functions for a long term benefit, so good idea.

Reply:
> wouldn't it only make sense with intersection types?
Hmm, maybe not:

```
> typechecker:collect_specs_types_opaques_and_functions([merl:quote("-spec f(a) -> ok; (b) -> nok.")]).
#parsedata{module = undefined,export_all = false,
           exports = [],imports = [],
           specs = [{{f,1},
                     [{type,1,'fun',[{type,1,product,[{atom,1,a}]},{atom,1,ok}]},
                      {type,1,'fun',
                            [{type,1,product,[{atom,1,b}]},{atom,1,nok}]}]}],
           types = [],opaques = [],records = [],functions = []}
```

I was thinking that at some point these return values get merged into a union, but having skimmed over the code now it doesn't seem to be the case. My bad.